File: brockman/helpers/Csv.rb
# encoding: utf-8

class Csv

  CACHE_SIZE = 500 # how many results to get
  FILE_BUFFER_SIZE = 32768 # 2 ^ 15 # how long the string will be before write to disk
  BASE_PATH = File.join( Dir.pwd, "csvs" )

  def initialize( options )

    @couch = options[:couch]
    @name  = options[:name]
    @path  = options[:path]
    @cachedResults = {}

  end


  # lazyish get from server
  def getResult( id )
    puts id
    # try to get it from the cache
    if @cachedResults[id].nil? # if the result isn't there get new ones

      nextResultIds = @orderedResults.slice!( 0, CACHE_SIZE )

      # fetch next results
      nextResults = @couch.postRequest({
        :view => "csvRows",
        :data => { "keys" => nextResultIds },
        :parseJson => true
      })

      # this doesn't keep the old results if there were any... @orderedResults important
      @cachedResults = Hash[nextResults['rows'].map { |row| [row['id'], row['value'] ] }]

    end

    result = @cachedResults[id]
    @cachedResults.delete(id)

  end

  def doWorkflow(options)

    resultsByTripId = options[:resultsByTripId]
    groupTimeZone = options[:groupTimeZone]

    machineNames = []
    columnNames  = []
    indexByMachineName = {}

    files = getFiles()

    # save all the result ids in order so we can can grab chunks
    @orderedResults = []
    resultsByTripId.each { | tripId, resultIds| @orderedResults.concat(resultIds) }

    # go through each trip and it's array of resultIds
    resultsByTripId.each { | tripId, resultIds |

      # make an array of resultIds for this trip
      results = resultIds.map { | resultId | getResult(resultId) }

      row = []

      results.each_with_index { | result, resultIndex |

        next if result.nil?

        for cell in result

          key         = cell['k']
          value       = cell['v']
          machineName = cell['m'] + resultIndex.to_s

          # hack for handling time
          isTimeRelated = key.match(/timestamp/) || key.match(/start_time/) || key.match(/startTime/)
          isntFalsy     = ! ( value.nil? || value == "" || value == 0 )
          if isTimeRelated && isntFalsy && groupTimeZone.nil?  then value = Time.at(value.to_i / 1000).strftime("%yy %mm %dd %Hh %Mm") end
          if isTimeRelated && isntFalsy && !groupTimeZone.nil? then value = Time.at(value.to_i / 1000).getlocal(groupTimeZone).strftime("%yy %mm %dd %Hh %Mm") end

          unless indexByMachineName[machineName] # Have we seen the machine name before?
            machineNames.push machineName
            columnNames.push key
            indexByMachineName[machineName] = machineNames.index(machineName)

          end

          index = indexByMachineName[machineName]

          row[index] = value

        end

      }

      files[:body].write row.map { |title| "\"#{title.to_s.gsub(/"/,'”')}\"" }.join(",") + "\n"

    }

    files[:header].write columnNames.map { |title| "\"#{title.to_s.gsub(/"/,'”')}\"" }.join(",") + "\n"

    files[:header].close()
    files[:body].close()

    `cat #{files[:headerUri]} #{files[:bodyUri]} > #{files[:fileUri]}`

    return { :uri => files[:fileUri], :name => files[:fileName] }

  end # of doWorkflow


  def doAssessment(options)

    resultIds = options[:resultIds]

    columnNames = []
    machineNames = []
    indexByMachineName = {}

    files = getFiles()

    # save all the result ids in order so we can can grab chunks
    @orderedResults = []
    resultIds.each { |value| @orderedResults.push value }

    resultIds.each { |resultId|

      row = []

      result = getResult(resultId)

      for cell in result

        key         = cell['k']
        value       = cell['v']
        machineName = cell['m']

        unless indexByMachineName[machineName] # Have we seen the machine name before?
          machineNames.push machineName
          columnNames.push key
          indexByMachineName[machineName] = machineNames.index(machineName)
        end

        index = indexByMachineName[machineName]

        row[index] = value

      end

      files[:body].write row.map { |value|
        value = value.to_s if value.class != String
        result = "\""
        (0..value.length-1).each { |index|
          result += if value[index] == '"' then '”' else value[index] end
        }
        result += "\""
      }.join(",") + "\n"

    }

    files[:header].write columnNames.map { |title| "\"#{title.to_s.gsub(/"/,'”')}\"" }.join(",") + "\n"

    files[:header].close()
    files[:body].close()
    `cat #{files[:headerUri]} #{files[:bodyUri]} > #{files[:fileUri]}`

    return { :uri => files[:fileUri], :name => files[:fileName] }

  end # of doAssessment

  private

  def getFiles()

    path = File.join( BASE_PATH, @path )

    # ensure group directory
    unless File.exists?( path )
      Dir.mkdir( path )
    end

    fileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv"
    fileUri  = File.join( path, fileName )

    bodyFileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv.body"
    bodyUri      = File.join( path, bodyFileName )
    bodyFile     = File.open( bodyUri, 'w' )

    headerFileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv.header"
    headerUri      = File.join( path, headerFileName )
    headerFile     = File.open( headerUri, 'w' )

    return {
      :header  => headerFile, :headerUri => headerUri,
      :body    => bodyFile,   :bodyUri   => bodyUri,
      :fileUri => fileUri,    :fileName  => fileName
    }

  end

end

File: csv-concept/csv.js
'use strict';

module.exports = {}

var log = console.log
var axios = require('axios')
var db = axios.create({
  baseURL: 'http://admin:password@localhost/db/group-national_tablet_program/',
  timeout: 999999000,
  headers: {'Content-Type': 'application/json'}
});

// A classic sleep function.
async function sleep(seconds) {
  return new Promise(function(resolve, reject) {
    let ms = seconds*1000
    setTimout(resolve, ms)
  })
}

module.exports.get = async function byWorkflowIdMonthYear(options) {

    let workflowId = options.workflowId
    let year = options.year
    let month = options.month

    // PORT: Start code from ./brockman/routes/workflow.rb
    var groupSettings = (await db.get('settings')).data
    let groupTimeZone = groupSettings['timeZone']    
    let locationList = (await db.get('location-list')).data 
    let workflow = (await db.get(workflowId)).data
    let workflowName = workflow['name']

    log(`Beginning ${workflowName}`)

    // Get results filtered by date where the value key in the view is the Trip ID.
    var response = await db.post(`/_design/ojai/_view/tutorTrips?reduce=false`, { 
      "keys": [`year${year}month${month}workflowId${workflowId}`],
    })
    let resultByDate = response.data.rows

    // Transform resultByDate into an object where keys are Trip ID and value is an array of Result IDs called resultsByTripId. 
    let resultsByTripId = {}
    let allTripIds = []
    resultByDate.forEach(function(row) {
      var tripId = row['value']
      resultsByTripId[tripId]  = (!resultsByTripId[tripId]) ? [] : resultsByTripId[tripId] 
      resultsByTripId[tripId].push(row['id'])
      allTripIds.push(tripId)
    })

    debugger

    // @TODO This looks unnessecary. resultsByTripId is already built.
    allTripIds.map(function(tripId) { 
      resultsByTripId[tripId]  = (!resultsByTripId[tripId]) ? [] : resultsByTripId[tripId] 
    })

    debugger

    // End ported code from ./brockman/routes/workflow.rb

    // Start ported code from ./brockman/helpers/Csv.rb#Csv.doWorkflow

    let machineNames = []
    let columnNames  = []
    let indexByMachineName = {}

    // @TODO Needs porting.
    //let files = await getFiles()

    let throttle = 0

    // save all the result ids in order so we can can grab chunks
    // @TODO I thik this is unnessesary. 
    this.orderedResults = []
    for (var tripId in resultsByTripId) {
      let resultIds = resultsByTripId[tripId] 
      this.orderedResults.concat(resultIds) 
    }

    // go through each trip and it's array of resultIds.
    for (var tripId in resultsByTripId) {
      let resultIds = resultsByTripId[tripId] 

      // make an array of resultIds for this trip
      // The above comment is off the mark. This makes an array of Result Document Objects for the Trip.
      // results = resultIds.map(function (resultId) { getResult(resultId) })
      // @TODO getResult using an in memory cache? Not so sure that helps much.
      var results = []
      for (let resultId of resultIds) { 
        var result = await (db.get(resultId)).data
        results.push(result)
      }

      debugger

      throttle = throttle + 1

      if ((throttle % 500) == 0) {
        log("Throttling...")
        await sleep(3)
      }

      let row = []

      for (resultsIndex = 0; resultsIndex < results.length; resultsIndex++) {
        let result = results[resultIndex]

        if (!result) continue;

        result.forEach(function(cell) { // result might be an Obj

          let key         = cell['k']
          let value       = cell['v']
          let machineName = cell['m'] + resultIndex.toString()

          // hack for handling time
          // PORT!
          let isTimeRelated = key.match(/timestamp/) || key.match(/start_time/) || key.match(/startTime/)
          /*
          isntFalsy     = ! ( value.nil? || value == "" || value == 0 )
          if isTimeRelated && isntFalsy && groupTimeZone.nil?  then value = Time.at(value.to_i / 1000).strftime("%yy %mm %dd %Hh %Mm %Ss") end
          if isTimeRelated && isntFalsy && !groupTimeZone.nil? then value = Time.at(value.to_i / 1000).getlocal(groupTimeZone).strftime("%yy %mm %dd %Hh %Mm %Ss") end

          unless indexByMachineName[machineName] # Have we seen the machine name before?
            machineNames.push machineName
            columnNames.push key
            indexByMachineName[machineName] = machineNames.index(machineName)

          end

          index = indexByMachineName[machineName]

          row[index] = value
          */

        })

      }

      //files[:body].write row.map { |title| "\"#{title.to_s.gsub(/"/,'”')}\"" }.join(",") + "\n"

    }
    /*
    files[:header].write columnNames.map { |title| "\"#{title.to_s.gsub(/"/,'”')}\"" }.join(",") + "\n"

    files[:header].close()
    files[:body].close()

    `cat #{files[:headerUri]} #{files[:bodyUri]} > #{files[:fileUri]}`

    return { :uri => files[:fileUri], :name => files[:fileName] }
    */
}

/*
  def getFiles()

    path = File.join( BASE_PATH, @path )

    # ensure group directory
    unless File.exists?( path )
      Dir.mkdir( path )
    end

    fileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv"
    fileUri  = File.join( path, fileName )

    bodyFileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv.body"
    bodyUri      = File.join( path, bodyFileName )
    bodyFile     = File.open( bodyUri, 'w' )

    headerFileName = @name.downcase().gsub( /[^a-zA-Z0-9]/, "-" ) + ".csv.header"
    headerUri      = File.join( path, headerFileName )
    headerFile     = File.open( headerUri, 'w' )

    return {
      :header  => headerFile, :headerUri => headerUri,
      :body    => bodyFile,   :bodyUri   => bodyUri,
      :fileUri => fileUri,    :fileName  => fileName
    }

  end
*/
